utils/zerovec/src/ule/vartuple.rs

```diff
@@ -0,0 +1,181 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+//! Types to help compose VarULE primitives.
+//!
+//! This module exports [`VarTuple`] and [`VarTupleULE`], which allow a single sized type and
+//! a single unsized type to be stored together as a [`VarULE`].
+//!
+//! # Examples
+//!
+//! ```
+//! use zerovec::ule::vartuple::{VarTuple, VarTupleULE};
+//! use zerovec::VarZeroVec;
+//!
+//! struct Employee<'a> {
+//!     id: u32,
+//!     name: &'a str,
+//! };
+//!
+//! let employees = [
+//!     Employee {
+//!         id: 12345,
+//!         name: "Jane Doe",
+//!     },
+//!     Employee {
+//!         id: 67890,
+//!         name: "John Doe",
+//!     },
+//! ];
+//!
+//! let employees_as_var_tuples = employees
+//!     .into_iter()
+//!     .map(|x| VarTuple {
+//!         sized: x.id,
+//!         variable: x.name,
+//!     })
+//!     .collect::<Vec<_>>();
+//!
+//! let employees_vzv: VarZeroVec<VarTupleULE<u32, str>> =
+//!     employees_as_var_tuples.as_slice().into();
+//!
+//! assert_eq!(employees_vzv.len(), 2);
+//!
+//! assert_eq!(employees_vzv.get(0).unwrap().sized.as_unsigned_int(), 12345);
+//! assert_eq!(&employees_vzv.get(0).unwrap().variable, "Jane Doe");
+//!
+//! assert_eq!(employees_vzv.get(1).unwrap().sized.as_unsigned_int(), 67890);
+//! assert_eq!(&employees_vzv.get(1).unwrap().variable, "John Doe");
+//! ```
+
+use core::mem::size_of;
+
+use super::{AsULE, EncodeAsVarULE, UleError, VarULE, ULE};
+
+/// A sized type that can be converted to a [`VarTupleULE`].
+///
+/// See the module for examples.
+#[derive(Debug)]
+#[allow(clippy::exhaustive_structs)] // well-defined type
+pub struct VarTuple<A, B> {
+    pub sized: A,
+    pub variable: B,
+}
+
+/// A dynamically-sized type combining a sized and an unsized type.
+///
+/// See the module for examples.
+#[derive(Debug)]
+#[allow(clippy::exhaustive_structs)] // well-defined type
+pub struct VarTupleULE<A: AsULE, V: ?Sized> {
```

Comment:
this needs to be repr(packed)

Reply:
Oh sure. `repr(C, packed)`, yes?

Reply:
yes

Reply:
It lets me apply `repr(C)`, but `repr(packed)` and `repr(C, packed)` say:

```
error[E0277]: the size for values of type `V` cannot be known at compilation time
  --> utils/zerovec/src/ule/vartuple.rs:74:19
   |
72 | pub struct VarTupleULE<A: AsULE, V: ?Sized> {
   |                                  - this type parameter needs to be `Sized`
73 |     pub sized: A::ULE,
74 |     pub variable: V,
   |                   ^ doesn't have a size known at compile-time
   |
   = note: the last field of a packed struct may only have a dynamically sized type if it does not need drop to be run
   = help: change the field's type to have a statically known size
```

Reply:
I think repr(C) is fine.

https://doc.rust-lang.org/reference/type-layout.html#the-c-representation

Our safety requirement is that we are align(1) with no padding.

The Rust Reference says:

> The alignment of the struct is the alignment of the most-aligned field in it.

Since the two generics are bound to `ULE` and `VarULE` which are defined to be align 1, then the repr(C) containing only those fields is defined to be align 1. (Note: I put those bounds on the impl, but I can move them to the struct to satisfy this)

> The size and offset of fields is determined by the following algorithm.
>
> Start with a current offset of 0 bytes.
>
> For each field in declaration order in the struct, first determine the size and alignment of the field. If the current offset is not a multiple of the field’s alignment, then add padding bytes to the current offset until it is a multiple of the field’s alignment. The offset for the field is what the current offset is now. Then increase the current offset by the size of the field.
>
> Finally, the size of the struct is the current offset rounded up to the nearest multiple of the struct’s alignment.

Since all field lengths are a multiple of 1, then repr(C) will never add padding bytes.